File: programs/ch02/perl/concord.pl
# Concordance program slightly modified from Doug Cooper. 
# In CORPORA Mailing list archive 
use utf8;
#binmode(STDOUT, ":encoding(UTF-8)");
#binmode(STDIN, ":encoding(UTF-8)");
#use open IO => ':utf8';

($file_name, $pattern, $width) = @ARGV;

open(FILE, "$file_name") || die "Could not open file $file_name.";

while ($line = <FILE>) {
	$text .= $line;
}
$pattern =~ s/ /\\s+/g; # Let spaces match across _and print_ newlines
$text =~ s/\s+/ /g; # Uncomment this to match/print newlines as spaces
while ($text =~ m/(.{0,$width}$pattern.{0,$width})/g ) {
	print "$1\n"; #$1 holds the match
	# print the string with 0..width
	# characters on either side
}

File: programs/ch02/python/concord.py
"""
Concordance program to find all the concordances
of a pattern surrounded by width characters.
Usage: python concord.py file pattern width
"""
__author__ = "Pierre Nugues"

import regex as re
import sys

[file_name, pattern, width] = sys.argv[1:]
try:
    text = open(file_name).read()
except:
    print('Could not open file', file_name)
    exit(0)

# spaces match tabs and newlines
pattern = re.sub(' ', '\\s+', pattern)
# Replaces newlines with spaces in the text
text = re.sub('\s+', ' ', text)
concordance = ('(.{{0,{width}}}{pattern}.{{0,{width}}})'
               .format(pattern=pattern, width=width))
for match in re.finditer(concordance, text):
    print(match.group(1))
# print the string with 0..width characters on either side
